clang/include/clang/Sema/Sema.h

```diff
@@ -12055,6 +12058,13 @@ class Sema final {
   // useful notes that shows where the kernel was called.
   bool ConstructingOpenCLKernel = false;
 
+  // We store map {call expr + its location} -> !isUnevaluatedContext for proper
```

Comment:
Just 50 cents from me. IMO this solution is a bit overkill.
This stuff is needed only because we diagnose anything from SemaSYCL and it's too late to use `isUnevaluatedContext()` function.
Have you considered reusing deferred diagnostics for your case?
First, it moves diagnosing from SemaSYCL to normal Sema, so you won't need this stuff at all.
Second, AFAIK other programming models also use it to validate calls (you know, in CUDA some functions for device, some for host, you can't call device function from host... and so on).
Basically, all you need is take a look at functions like `CheckCUDACall` and `checkOpenMPDeviceFunction` (and their usages) and add some logic to our version of such function which is called `checkSYCLDeviceFunction` (and some usages of it into SEMA).

Reply:
Done. You may now take a look at the changes.